Introduce RoutePattern

rynowak opened the pull request:
Introduces RoutePattern - a new parser and representation for routing
templates, defaults, and constraints.

This is a new representation for all of the 'inputs' to routing that is
immutable and captures 'out of line' information for defaults and
constraints.

This will allow us to unify the handling of constraints and values from
attribute style routes and conventional style routes.

rynowak commented:
/cc @kichalla and @JamesNK  - please take a look

The topic of how to deal with 'defaults' and constraints that are specified out of line keeps coming up in discussions 😆. It seemed worth it to me to introduce an object model that can represent all of these concepts.

The bulk of the this work was done during 2.1 and I'm just porting it forward because it seems like we'd benefit from it.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs:
```diff
@@ -0,0 +1,245 @@
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    internal static class RouteParameterParser
```
This is not a great place to start reviewing. This code is a virtual cut/paste from existing code - and some of it is total spaghetti (not my doing lol).

I relayered the existing parser stuff to use the new stuff.

rynowak replied:
I don't really think so since we ultimately need to surface all of the substrings we create as `string`. Also this code runs at startup, if this is on the critical path for anyone then we haven't done our job 😆

rynowak replied:
> if this is on the critical path for anyone then we haven't done our job

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePattern.cs:
```diff
@@ -0,0 +1,94 @@
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    [DebuggerDisplay("{DebuggerToString()}")]
+    public sealed class RoutePattern
```
This is a good place to start. There's a set of `RoutePattern*` classes introduced by this PR that make up the new object model.

davidfowl replied:
If we're going to make a new thing public, we should make sure it's using all the right types (Span etc).

JamesNK replied:
Route patterns (templates) are created once on startup and then live for the lifetime of the app. I don't see Span providing any significant benefit.

RoutePattern should expose whatever is the fastest for matches to be made with. I'm guessing that is `string` but @rynowak has been working on the optimizing matching and would know best.

davidfowl replied:
The fact that it's a public API makes me think about usages outside of ours (once at startup).

rynowak replied:
this code plays no role in URL matching

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,698 @@
+{
+    public static class RoutePatternFactory
+    {
+        public static RoutePattern Pattern(IEnumerable<RoutePatternPathSegment> segments)
```
Hmm, my first thought is I'd lean towards having `static Create(...)` methods on the various types, e.g. `var myNewRoutePattern = RoutePattern.Create(segements)`, instead of a central static help class

rynowak replied:
The code we wrote for this in 2.1 was like this and upon review I thought it was horrible because you end up needing to know like 8 different types. If you're using these then you'd want to `static using` this class and it's pretty nice/simple (see the tests).

kichalla replied:
I would prefer having having the construction of RoutePattern in one place, in this case RoutePatternFactory where I supply 'rawtext and non-inline defaults and constraints' and removing the creation methods on `RoutePattern` class. Some of these methods can be internal I think.

rynowak replied:
If you guys are really sad about this btw we can make it internal. 

@kichalla  if I understand your suggestion are you asking for me to move the `Parse` methods to this class?

kichalla replied:
> @kichalla if I understand your suggestion are you asking for me to move the Parse methods to this class?

Yes @rynowak

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParameterPart.cs:
```diff
@@ -0,0 +1,45 @@
+
+        internal override string DebuggerToString()
+        {
+            return RawText ?? "{" + (IsCatchAll ? "*" : string.Empty) + Name + (IsOptional ? "?" : string.Empty) + "}";
```
Since you're here, you could copy and paste this logic: https://github.com/aspnet/Mvc/blob/master/src/Microsoft.AspNetCore.Mvc.Core/Internal/RouteTemplateWriter.cs

rynowak replied:
Sure 👍

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternPart.cs:
```diff
@@ -0,0 +1,31 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+
```
nit: too liney

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternPathSegment.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public IReadOnlyList<RoutePatternPart> Parts { get; }
+
+        public string RawText { get; set; }
```
Should this have a public setter?

rynowak replied:
Good catch, this should not be settable.

JamesNK commented on src/Microsoft.AspNetCore.Routing/Template/TemplateSegment.cs:
```diff
@@ -4,12 +4,22 @@
+
         public bool IsSimple => Parts.Count == 1;
 
         public List<TemplatePart> Parts { get; } = new List<TemplatePart>();
```
Move this to the default `ctor()`? Avoids it being allocated twice in the other ctor.

rynowak replied:
Sure

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternException.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    public class RoutePatternException : Exception
```
👍

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePattern.cs:
```diff
@@ -0,0 +1,94 @@
+            }
+            catch (RoutePatternException ex)
+            {
+                throw new ArgumentException(ex.Message, nameof(pattern), ex);
```
Not sure really, I wrote this a long time ago. I'll undo this and see what happens.

kichalla commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternConstraintReference.cs:
```diff
@@ -0,0 +1,47 @@
+    [DebuggerDisplay("{DebuggerToString()}")]
+    public sealed class RoutePatternConstraintReference
+    {
+        internal RoutePatternConstraintReference(string rawText, string name, string content)
```
Assign name to Name property. Could you also add a sample in a comment here so that its clear (mainly how `rawText` and `content` differ is not clear for me)

kichalla replied:
Nit: name => parameterName

rynowak replied:
I've been using the property/variable name `name` for parameter name. I can change it here but I will also change it in a few other places. @JamesNK @kichalla thoughts?

Also probably `text` vs `rawText`

kichalla replied:
> I've been using the property/variable name name for parameter name. I can change it here but I will also change it in a few other places. @JamesNK @kichalla thoughts?

Sure 👍

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs:
```diff
@@ -0,0 +1,573 @@
+            var templatePart = RouteParameterParser.ParseRouteParameter(text, decoded);
+
+            // See #475 - this is here because InlineRouteParameterParser can't return errors
+            if (decoded.StartsWith("*") && decoded.EndsWith("?"))
```
Thanks. Did a search and fixed a few locations.

JamesNK replied:
`decoded.Length > 0 && decoded[0] == '*' && decoded[decoded.Length - 1] == '?'`?

Will probably be slightly faster.

rynowak commented:
Updated YO

JamesNK commented:
Should this be targeting 2.2 branch?

davidfowl commented on src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs:
```diff
@@ -0,0 +1,246 @@
+        //
+        // Ex: {hello} - this method is responsible for parsing 'hello'
+        // The factoring between this class and RoutePatternParser is due to legacy.
+        public static RoutePatternParameterPart ParseRouteParameter(string text, string parameter)
```
Got benchmarks for this?

davidfowl commented on src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs:
```diff
@@ -0,0 +1,246 @@
+                    // Parameter names are allowed to start with delimiters used to denote constraints or default values.
+                    // i.e. "=foo" or ":bar" would be treated as parameter names rather than default value or constraint
+                    // specifications.
+                    parameterName = parameter.Substring(startIndex, currentIndex - startIndex);
```
Any benefit using StringSegment here?

rynowak commented:
The code I'm adding here runs at startup to parse route templates. We've had infrastructure like this for a few years now without it coming up as perf concern, so I'm ignoring all perf-related feedback.

davidfowl commented:
> We've had infrastructure like this for a few years now without it coming up as perf concern, so I'm ignoring all perf-related feedback.

I'm not sure why. Then why make this a new top level public low level route parsing API? Why not internal if we only care about this one use case?

rynowak commented:
This code parses and validates the template, it plays no role in matching URLs.

rynowak commented:
I've updated this with @kichalla naming feedback and changed the base branch (thanks @JamesNK).

Regarding the (*other*) feedback. It seems like the main point of discussion here is where the object model (`RoutePattern`, `RoutePatternPathSegment`, `RoutePatternParameterPart`) expose spans or strings. If you're giving this feedback please take a look at these classes and let me know if you think span belongs here, and what benefits that will add. Is span a good fit for these APIs?

If you think that Span does belong here, then let me know if removing 'RawText` from all of the various pieces would change your mind.

kichalla commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternConstraintReference.cs:
```diff
@@ -0,0 +1,47 @@
+    [DebuggerDisplay("{DebuggerToString()}")]
+    public sealed class RoutePatternConstraintReference
+    {
+        internal RoutePatternConstraintReference(string rawText, string parameterName, string content)
```
Set the parameterName to the property

kichalla commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternConstraintReference.cs:
```diff
@@ -0,0 +1,47 @@
+            Content = content;
+        }
+
+        internal RoutePatternConstraintReference(string name, IRouteConstraint constraint)
```
nit: parameterName

rynowak commented:
@JamesNK @davidfowl - do you have additional thoughts about [this](https://github.com/aspnet/Routing/pull/585#issuecomment-403899783)?

JamesNK commented:
My original comment about using Span was about eliminating any intermediary strings. I like exposing String publicly.

rynowak commented:
> My original comment about using Span was about eliminating any intermediary strings. I like exposing String publicly. 

OK cool. If we find that the perf of the parsing matters, we could always retrofit the parsing code to use spans for fewer substrings.

I think I'll also take the step of removing `RawText` from most of the APIs (I'll leave it on route pattern).  There's no code that actually consumes rawtext other than on RoutePattern and it would make it more beneficial to use span if we need to later.

rynowak commented:
@JamesNK - any other feedback about this?

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+
+        private static RoutePatternPathSegment SegmentCore(IEnumerable<RoutePatternPart> parts)
+        {
+            return new RoutePatternPathSegment(parts.ToArray());
```
`parts` could sometimes already be an array (I don't know how deep you want to go on perf at the moment)

rynowak replied:
I'd still want to make a defensive copy if the user called `Segment(myArray)`

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+            return SeparatorPartCore(content);
+        }
+
+        private static RoutePatternSeparatorPart SeparatorPartCore( string content)
```
nit: extra space

rynowak replied:
heh

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+
+        private static RoutePatternSeparatorPart SeparatorPartCore( string content)
+        {
+            return new RoutePatternSeparatorPart( content);
```
nit: extra space

rynowak replied:
wierd

rynowak replied:
I wonder if the editor did that when I removed the `rawText` parameter

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+
+            if (parameterName.IndexOfAny(RoutePatternParser.InvalidParameterNameChars) >= 0)
+            {
+                throw new ArgumentException(Resources.FormatTemplateRoute_InvalidParameterName(parameterName));
```
`if (string.IsNullOrEmpty(parameterName))` and `if (parameterName.IndexOfAny(RoutePatternParser.InvalidParameterNameChars) >= 0)` are checked in `ParameterPartCore` and all its `ParameterPart` callers.

Don't need to duplicate check.

rynowak replied:
We generally put all argument validation as close as possible to the method the user called (see null checks all over the codebase).

JamesNK replied:
In that case you could leave this check (and in other overloads) and remove the one in ParameterPartCore.

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs:
```diff
@@ -0,0 +1,511 @@
+            }
+
+            RoutePattern = pattern;
+            Defaults = defaults ?? new RouteValueDictionary();
```
Allowing a null defaults feels weird.

Maybe a ctor that takes just pattern and another that takes both and checks defaults to not be null?

rynowak replied:
See your comment else where about `Exception(string message)`. It's normal to accept null for a parameter that's optional. 

I think the methods in MVC for `MapRoute` also will tolerate null, so it would require extra work all over the code 😆

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs:
```diff
@@ -0,0 +1,573 @@
+
+                if (context.Index <= i)
+                {
+                    throw new InvalidProgramException("Infinite loop in the parser. This is a bug.");
```
I don't like "This is a bug." in error message to user

rynowak replied:
What would you suggest? This is not a message a user should be able to see. I want this code to crash hard if there's an infinite loop.

JamesNK replied:
I would have something like "An unexpected infinite loop detected in the parser." and then a comment saying we should never reach here

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs:
```diff
@@ -0,0 +1,573 @@
+
+                if (context.Index <= i)
+                {
+                    throw new InvalidProgramException("Infinite loop in the parser. This is a bug.");
```
^

rynowak replied:
Again, what do you suggest?

JamesNK replied:
^

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs:
```diff
@@ -0,0 +1,573 @@
+            private int _index;
+            private int? _mark;
+
+            private HashSet<string> _parameterNames = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
readonly

rynowak replied:
fix't

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs:
```diff
@@ -0,0 +1,573 @@
+            {
+                var part = parts[i];
+
+                if (part.IsParameter && ((RoutePatternParameterPart)part).IsOptional && parts.Count > 1)
```
Part is cast to `RoutePatternParameterPart` many times in the code below. Create a cast variable?

rynowak replied:
This can probably be simplified now that we have pattern patching.

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternPart.cs:
```diff
@@ -0,0 +1,27 @@
+        // needs to be aware of what kind of parts we support.
+        //
+        // It is abstract so we can add semantics later inside the library.
+        internal RoutePatternPart(RoutePatternPartKind partKind)
```
Did the private AND internal keyword ever get added to C#? A candidate to use it right here 😝

rynowak replied:
C# 7.2! DOPE

JamesNK commented on src/Microsoft.AspNetCore.Routing/Template/TemplatePart.cs:
```diff
@@ -5,12 +5,47 @@
+            else
+            {
+                // Unreachable
+                throw new NotSupportedException();
```
Is this an ok thing to do or was it done in a hurry?

rynowak replied:
It should be unreachable code. What would you suggest?

JamesNK replied:
If you're fine with it, I'm fine with it. Just checking

JamesNK commented on test/Microsoft.AspNetCore.Routing.Tests/Template/TemplateParserTests.cs:
```diff
@@ -528,8 +525,8 @@ public void Parse_RegularExpressions_Unescaped(string template)
-        [InlineData("{p1?}{p2}", "p1", "p2")]
-        [InlineData("{p1?}{p2?}", "p1", "p2")]
+        [InlineData("{p1?}{p2}", "p1", "{p2}")]
+        [InlineData("{p1?}{p2?}", "p1", "{p2?}")]
```
How come you changed these?

rynowak replied:
I changed what gets passed to the error messages. now instead of just including the parameter name or an unrelated thing, it includes the text of the next part..

JamesNK commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePattern.cs:
```diff
@@ -0,0 +1,69 @@
+            PathSegments = pathSegments;
+        }
+
+        public IReadOnlyDictionary<string, object> Defaults { get; }
```
And collections on other patterns classes

rynowak replied:
I'm not concerned at all. To me that's equivalent to doing private reflection.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs:
```diff
@@ -0,0 +1,246 @@
+                    case ParseState.Start:
+                        switch (currentChar)
+                        {
+                            case null:
```
I'm really really unconcerned about perf in this code until we have a scenario that dictates that it matters. Each route template is parsed once at startup and never again.

rynowak replied:
Another meta comment about this. All of this code is rename from existing code. I don't mind addressing minor feedback but if we were going to spend time here I would throw it away and start over. I don't think rewriting it is anywhere near a priority.

This code was written in a pretty crazy way by someone who was sulking about being made to redo it after they were told not to use a regex for this. Seriously.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs:
```diff
@@ -0,0 +1,246 @@
+        {
+            public int CurrentIndex;
+
+            public IEnumerable<RoutePatternConstraintReference> Constraints;
```
sure

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RouteParameterParser.cs:
```diff
@@ -0,0 +1,246 @@
+        //
+        // Ex: {hello} - this method is responsible for parsing 'hello'
+        // The factoring between this class and RoutePatternParser is due to legacy.
+        public static RoutePatternParameterPart ParseRouteParameter(string text, string parameter)
```
cool beans 👍

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePattern.cs:
```diff
@@ -0,0 +1,69 @@
+        /// </summary>
+        /// <param name="name">The name of the parameter to match.</param>
+        /// <returns>The matching parameter or <c>null</c> if no parameter matches the given name.</returns>
+        public RoutePatternParameterPart GetParameter(string name)
```
Sure

rynowak replied:
This is fixed btw, just not showing on in the diff.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternException.cs:
```diff
@@ -0,0 +1,42 @@
+                throw new ArgumentNullException(nameof(pattern));
+            }
+
+            if (message == null)
```
Makes sense. Usually if something is optional, then it will tolerate null.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+                foreach (var kvp in constraints)
+                {
+                    updatedConstraints.Add(kvp.Key, new List<RoutePatternConstraintReference>()
+                {
```
sure

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+            for (var i = 0; i < updatedSegments.Length; i++)
+            {
+                var segment = updatedSegments[i];
+                for (var j = 0; j < segment.Parts.Count; j++)
```
seems fine

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs:
```diff
@@ -0,0 +1,517 @@
+            return new RoutePattern(
+                rawText,
+                updatedDefaults,
+                updatedConstraints.ToDictionary(kvp => kvp.Key, kvp => (IReadOnlyList<RoutePatternConstraintReference>)kvp.Value.ToArray()),
```
I don't have a scenario where this code is perf-sensitive.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternMatcher.cs:
```diff
@@ -0,0 +1,511 @@
+
+        private static readonly char[] Delimiters = new char[] { SeparatorChar };
+
+        public RoutePatternMatcher(
```
FYI this class is just a rename of existing code with a few details change to use `RoutePattern`.

rynowak commented on src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternParser.cs:
```diff
@@ -0,0 +1,573 @@
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    internal static class RoutePatternParser
```
This is also another rename of an existing class, the changes here are to support the `RoutePattern` instead of `RouteTemplate`.

rynowak commented:
Updated and addressed comments so far
